components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/AccessTokenIssuer.java

```diff
@@ -264,6 +264,8 @@ public OAuth2AccessTokenRespDTO issue(OAuth2AccessTokenReqDTO tokenReqDTO)
                 log.debug("Invalid scope provided by client Id: " + tokenReqDTO.getClientId());
             }
             tokenRespDTO = handleError(OAuthError.TokenResponse.INVALID_SCOPE, "Invalid Scope!", tokenReqDTO);
+            // Set custom error message if it is availabe in tokReqMsgCtx
+            setResponseErrorMessage(tokReqMsgCtx, tokenRespDTO);
```

Comment:
@YanyanZeng Thanks for the contribution. As we can see in your fix, It is specific to invalid scope scenario. But if we are to do such improvement we need to do it in a generic way where it can be used to customize the error message in all the relevant places.

Reply:
We are using the handleError() method[1] to build the error message. This method takes in errorCode, errorMessage and TokenRequestDTO object as argument.

What if we make this method protected and change the signature to,
```
protected OAuth2AccessTokenRespDTO handleError(String errorCode, String errorMsg,	                                                 OAuthTokenReqMessageContext tokReqMsgCtx)                                                
```

And implement the change suggested by @YanyanZeng. ie. We check whether a custom errorCode or errorMessage is set in the OAuthTokenReqMessageContext object. If so, we use it to build the error response. Else we go with the default error code and error message passed to the method.






[1] https://github.com/wso2-extensions/identity-inbound-auth-oauth/blob/8675a5b6c51ddf1e39724acf5fff6488e631d157/components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/AccessTokenIssuer.java#L443